invenio_base/urls/builder.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2025 CERN.
+# Copyright (C) 2025 Northwestern University.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""InvenioUrlsBuilder.
+
+This code is in support of providing `invenio_url_for`, the RDM supplement
+to Flask's `url_for` which allows any part of the application (in the broad
+sense of the term) to generate URLs even if those URLs are for views registered
+in another Flask application. Other developer niceties are included.
+"""
+
+from abc import ABC, abstractmethod
+
+from flask import Flask, current_app
+from werkzeug.routing import BuildError, Map, Rule
+
+from invenio_base.app import blueprint_loader
+
+
+class InvenioUrlsBuilder(ABC):
+    """Interface of class in charge of producing urls."""
+
+    @abstractmethod
+    def build(self, endpoint, values, method=None):
+        """Build current or other app url."""
+
+
+class InvenioAppsUrlsBuilder(InvenioUrlsBuilder):
```

Comment:
To be honest, this is more of a `contrib` category of code as it implements some of the interface mechanisms defined in invenio-base for specific usage in invenio-app. I still think it's fine to place here, as `create_app_factory` and `wsgi_factory` are also kind of created for specific usage in invenio-app. I am also of the philosophy that it's also much easier in the development phase to stick everything together at first and splinter later.